src/common/cups_print.c

```diff
@@ -0,0 +1,538 @@
+}
+
+void dt_print_file(const int32_t imgid, const char *filename, const dt_print_info_t *pinfo)
+{
+  // first for safety check that filename exists and is readable
+
+  if (!g_file_test(filename, G_FILE_TEST_IS_REGULAR))
+  {
+    dt_control_log(_("file '%s' to print not found for image `%d' on %s"), filename, imgid, pinfo->printer.name);
+    return;
+  }
+
+  cups_dest_t *dests;
+  int num_dests = cupsGetDests(&dests);
+  cups_dest_t *dest = cupsGetDest(pinfo->printer.name, NULL, num_dests, dests);
+
+  cups_option_t *options = NULL;
+  int num_options = 0;
+
+  for (int j = 0; j < dest->num_options; j ++)
+    if (cupsGetOption(dest->options[j].name, num_options,
+                      options) == NULL)
+      num_options = cupsAddOption(dest->options[j].name,
+                                  dest->options[j].value,
+                                  num_options, &options);
+
+  // disable cm on CUPS, this is important as dt does the cm
+
+  if (*pinfo->printer.profile)
+    num_options = cupsAddOption("cm-calibration", "true", num_options, &options);
+
+  // media to print on
+
+  num_options = cupsAddOption("media", pinfo->paper.name, num_options, &options);
+
+  // never print two-side
+
+  num_options = cupsAddOption("sides", "one-sided", num_options, &options);
+
+  // and a single image per page
+
+  num_options = cupsAddOption("number-up", "1", num_options, &options);
+
+  // if the printer has no hardward margins activate the borderless mode
+
+  if (pinfo->printer.hw_margin_top == 0 || pinfo->printer.hw_margin_bottom == 0
+      || pinfo->printer.hw_margin_left == 0 || pinfo->printer.hw_margin_right == 0)
+  {
+    // there is many vaariant for this parameter
+    num_options = cupsAddOption("StpFullBleed", "true", num_options, &options);
+    num_options = cupsAddOption("STP_FullBleed", "true", num_options, &options);
+    num_options = cupsAddOption("Borderless", "true", num_options, &options);
+  }
+
+  if (pinfo->page.landscape)
+    num_options = cupsAddOption("landscape", "true", num_options, &options);
+  else
+    num_options = cupsAddOption("landscape", "false", num_options, &options);
+
+  // print lp options
+
+  if (darktable.unmuted & DT_DEBUG_PRINT)
+  {
+    printf ("[print] printer options (%d)\n", num_options);
+    for (int k=0; k<num_options; k++)
+    {
+      printf ("[print]   %s=%s\n", options[k].name, options[k].value);
+    }
+  }
+
+  const int job_id = cupsPrintFile(pinfo->printer.name, filename,  "darktable", num_options, options);
+
+  if (job_id == 0)
+    dt_control_log(_("error while printing image %d on `%s'"), imgid, pinfo->printer.name);
+  else
+    dt_control_log(_("printing image %d on `%s'"), imgid, pinfo->printer.name);
+
+  cupsFreeOptions (num_options, options);
+}
+
+static void _get_image_dimension (int32_t imgid, int32_t *iwidth, int32_t *iheight)
+{
+  dt_mipmap_buffer_t buf;
+  dt_mipmap_cache_get(darktable.mipmap_cache, &buf, imgid, DT_MIPMAP_3, DT_MIPMAP_BEST_EFFORT, 'r');
```

Comment:
This is not going to work, mipmap 3 has a fixed size that you can't get the image size from.